heron/packing/src/java/com/twitter/heron/packing/builder/PackingPlanBuilder.java

```diff
@@ -107,13 +112,80 @@ public PackingPlanBuilder addInstance(Integer containerId,
     return this;
   }
 
+  @SuppressWarnings("JavadocMethod")
+  /**
+   * Add an instance to the first container possible ranked by score.
+   * @return containerId of the container the instance was added to
+   * @throws com.twitter.heron.packing.ResourceExceededException if the instance could not be added
+   */
+  public int addInstance(Scorer<Container> scorer,
+                         InstanceId instanceId) throws ResourceExceededException {
+    List<Scorer<Container>> scorers = new LinkedList<>();
+    scorers.add(scorer);
+    return addInstance(scorers, instanceId);
+  }
+
+  @SuppressWarnings("JavadocMethod")
+  /**
+   * Add an instance to the first container possible ranked by score. If a scoring tie exists,
+   * uses the next scorer in the scorers list to break the tie.
+   * @return containerId of the container the instance was added to
+   * @throws com.twitter.heron.packing.ResourceExceededException if the instance could not be added
+   */
+  private int addInstance(List<Scorer<Container>> scorers, InstanceId instanceId)
+      throws ResourceExceededException {
+    initContainers();
+    for (Container container : sortContainers(scorers, this.containers.values())) {
+      try {
+        addInstance(container.getContainerId(), instanceId);
+        return container.getContainerId();
+      } catch (ResourceExceededException e) {
+        // ignore since we'll continue trying
+      }
+    }
+    //Not enough containers.
+    throw new ResourceExceededException(String.format(
+        "Insufficient resources to add instance %s to any of the %d containers.",
+        instanceId, this.containers.size()));
+  }
+
   public boolean removeInstance(Integer containerId, String componentName) {
     initContainers();
     Optional<PackingPlan.InstancePlan> instancePlan =
         containers.get(containerId).removeAnyInstanceOfComponent(componentName);
     return instancePlan.isPresent();
   }
 
+  @SuppressWarnings("JavadocMethod")
+  /**
+   * Remove an instance from the first container possible ranked by score.
+   * @return containerId of the container the instance was removed from
+   * @throws com.twitter.heron.spi.packing.PackingException if the instance could not be removed
+   */
+  public int removeInstance(Scorer<Container> scorer, String componentName) {
+    List<Scorer<Container>> scorers = new LinkedList<>();
+    scorers.add(scorer);
+    return removeInstance(scorers, componentName);
+  }
+
+  @SuppressWarnings("JavadocMethod")
+  /**
+   * Remove an instance from the first container possible ranked by score. If a scoring tie exists,
+   * uses the next scorer in the scorers list to break the tie.
+   * @return containerId of the container the instance was removed from
+   * @throws com.twitter.heron.spi.packing.PackingException if the instance could not be removed
+   */
+  public int removeInstance(List<Scorer<Container>> scorers, String componentName) {
```

Comment:
In the RCRR implementation, during scale down we do not always start from the first container. For example if we need to scale down 2 components by 1 instance from containers [1...5] and the first component is removed from instance 2 then the removal for the next component will start from container 3. This will help with load balancing. This remove instance method does not capture that. However, it is fine if a good scoring method is used when we change the scale down approach as discussed in the other thread.